internal/server/users.go

```diff
@@ -0,0 +1,155 @@
+package server
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	"golang.org/x/crypto/bcrypt"
+	validator "gopkg.in/go-playground/validator.v9"
+
+	ihttp "github.com/Pigmice2733/peregrine-backend/internal/http"
+	"github.com/Pigmice2733/peregrine-backend/internal/store"
+	jwt "github.com/dgrijalva/jwt-go"
+)
+
+type contextKey string
+
+const (
+	keyRolesContext   contextKey = "pigmice_roles"
+	keySubjectContext contextKey = "pigmice_subject"
+)
+
+type baseUser struct {
+	Username string `json:"username" validate:"gte=4,lte=32"`
+	Password string `json:"password" validate:"gte=8,lte=128"`
+}
+
+func (s *Server) authenticateHandler() http.HandlerFunc {
```

Comment:
Is there a way to force connections to `authenticate` or any authenticated route to use SSL? Or would the server just not accept the request but the user would still have managed to spill their plaintext credentials all over the place?

Reply:
They deserve it if they don't use SSL. And yeah they would have still sent it so it wouldn't matter. Maybe the frontend could do something with that but I think it's fine.